### ISSUE TITLE
Add dependency types option

### ISSUE THREAD

**Issue by DIDoS** [OSSIndex/auditjs#18]
Currently there is no possibility I know of to select which [dependencies ](https://docs.npmjs.com/files/package.json#dependencies) (devDependencies, dependencies, peerDependencies, bundledDependencies, optionalDependencies) are included to search for vulnerabilities.

An commandline option/argument like "dependencyTypes=" or --prod-only or anything to change the kind of dependencies taken would be appreciated.

**Comment by OSSIndex-Admin** [comment]
I like this idea. Things are crazy busy at the moment, but I will try and get to this as soon as possible.

**Comment by OSSIndex-Admin** [comment]
Sorry it took so long. I added the requested options. It looks like this:

```
node audit.js -p test_data/package.json --dependencyTypes devDependencies,dependencies,peerDependencies,bundledDependencies,optionalDependencies
```

`-d` may be used instead of `--dependencyTypes`

Note the argument is a comma separated list of dependency types.

**Comment by thom-nic** [comment]
It seems this option only works in conjunction with the `-p` option, meaning it only works on top-level dependencies.  Might it be possible to enhance this capability such that 
```
auditjs -d dependencies
```
scans the entire production dependency tree? Calling `npm ls --production` gives the whole production dependency tree so hopefully this is possible for auditjs to achieve.

For example, if I run `auditjs` without any dependency type filtering, `marked v0.3.6` (dependency to `jsdoc` and `minimatch` (`gulp` dep) are flagged even though jsdoc and gulp are devDependencies.  Happy to open a new issue if you're willing to consider this enhancement.

**Comment by OSSIndex-Admin** [comment]
You make a very good point. I will look into the interfaces and see if I can expand the feature some.

**Comment by OSSIndex-Admin** [comment]
Alright. I finally have a few seconds to rub together, and have been exploring a few ideas. It seems that the examples you have provided have a bit of an issue:

* From what I can tell `marked` is a production dependency of `jsdoc`
* It seems that for `gulp` that `minimist` is a production dependency, but I cannot find `minimatch` anywhere.

The reason this is important is that the way we are currently using the npm.command.ls API we should be only getting production dependencies and **always** ignoring dev dependencies. The code in "audit.js" runs [npm.command.ls](https://github.com/OSSIndex/auditjs/blob/master/audit.js#L144) with no arguments and parses the full results. The parsing code only looks at the [dependency](https://github.com/OSSIndex/auditjs/blob/master/audit.js#L352) field in the results, while the devDependency field is ignored.

I can, of course, add support for monitoring development dependencies as well, but the default behaviour should be production.

I will do some more testing over the next couple of days to ensure I am not wrong, Please do let me know if there is a flaw in my reasoning.

**Comment by thom-nic** [comment]
Ok I went back and looked at this again... `minimatch` was a bad example, but `marked` is def showing up when it should not.  Here's an example:

```
$ auditjs --version
2.2.2

$ auditjs -d dependencies|grep marked
[401/873] marked 0.3.6  [VULNERABLE]   
  * https://github.com/chjj/marked/commit/8f9d0b72f5606ed32057049f387161dd41c36ade
  * https://github.com/chjj/marked/commit/cd2f6f5b7091154c5526e79b5f3bfb4d15995a51
  * https://github.com/chjj/marked/issues/844
  * https://github.com/chjj/marked/issues/848

$ npm ls --prod marked
myproj@1.0.0 /Users/blabhablahb/myproj
└── (empty)


$ npm ls marked       
myproj@1.0.0 /Users/blabhablahb/myproj
└─┬ jsdoc@3.4.3
  └── marked@0.3.6 
```

**Comment by ken-duck** [comment]
The issue has been reproduced locally. I am working on resolving the issue today.

```
bragi:issue-18 twoducks$ auditjs -d dependencies|grep marked
[14/20] marked 0.3.6  [VULNERABLE]   
  * https://github.com/chjj/marked/commit/8f9d0b72f5606ed32057049f387161dd41c36ade
  * https://github.com/chjj/marked/commit/cd2f6f5b7091154c5526e79b5f3bfb4d15995a51
  * https://github.com/chjj/marked/issues/844
  * https://github.com/chjj/marked/issues/848
bragi:issue-18 twoducks$ npm ls marked
issue-18@2.2.2 /Users/twoducks/git.vor/auditjs/issue-18
└─┬ jsdoc@3.4.3
  └── marked@0.3.6 

bragi:issue-18 twoducks$ npm ls --prod marked
issue-18@2.2.2 /Users/twoducks/git.vor/auditjs/issue-18
└── (empty)
```

**Comment by ken-duck** [comment]
I have added a `--prod` or `--production` option which will only output results for production dependencies. I am going to unify this with the `--dependencyTypes` option so that both command line options work in both full-project and package file invocations of the tool before pushing to npm.

**Comment by ken-duck** [comment]
Alright, the new version has been published and should be available soon. You have these options, all of which have the same effect:

```
auditjs --prod
auditjs --production
auditjs -d dependencies
auditjs -dependencyTypes dependencies
```

**Comment by ken-duck** [comment]
(You are looking for version 2.2.3)

**Comment by ken-duck** [comment]
The best version so far is now 2.2.4, but there remains a bug whereupon some dependencies are lost when run in --prod mode. I know the reason, but the fix is a tad more complicated. I am leaving this open for now...

**Comment by ken-duck** [comment]
Attempt number 2. I have published a new version (2.2.5) which hopefully resolves the new problems I found and properly resolves the dependency list as requested. Keep an eye open for missing dependencies (all my manual tests work so far, I need to work on adding much more rigorous automated tests).
